File: lab1/materie.cpp
#include "materie.h"
#include "facultate.h"
#include "student.h"
#include <string>
using namespace std;

Materie::Materie(int id, int anDeStudiu, string nume, int credite,
      unordered_map<int, Profesor*> titulariCurs, Facultate *facultate) {
   this->id = id;
   this->anDeStudiu = anDeStudiu;
   this->nume = nume;
   this->credite = credite;
   this->titulariCurs = titulariCurs;

   for (const auto &student : facultate->studentiAn[anDeStudiu]) {
      this->note[student->nrMatricol] = unordered_map<string, float>();
   }
   facultate->materii.push_back(this);
}

File: lab1_java/src/main/java/com/lab1_java/Materie.java
package com.lab1_java;

import java.util.HashMap;

public abstract class Materie {
    public int id;
    public int anDeStudiu;
    public String nume;
    public int credite;
    public HashMap<Integer, Profesor> titulariCurs;
    public HashMap<String, HashMap<String, Float>> note; // <studentMatricol, tipnota, nota>
    public HashMap<String, Float> medii;
    
    Materie(int id, int anDeStudiu, String nume, int credite, HashMap<Integer, Profesor> titulariCurs, Facultate facultate) {
        this.id = id;
        this.anDeStudiu = anDeStudiu;
        this.nume = nume;
        this.credite = credite;
        this.titulariCurs = titulariCurs;
        this.note = new HashMap<>();
        this.medii = new HashMap<>();
        
        for (Student student : facultate.studentiAn.get(anDeStudiu)) {
            this.note.put(student.nrMatricol, new HashMap<>());
        }
    }
    
    public abstract float evaluare(String studentMatricol);
}
